coalib/settings/SectionFilling.py

```diff
@@ -25,17 +26,38 @@ def fill_settings(sections, acquire_settings, log_printer):
     local_bears = {}
     global_bears = {}
 
+    bears_by_section = {}
+    bear_dirs_by_section = {}
+
     for section_name, section in sections.items():
-        bear_dirs = section.bear_dirs()
-        bears = list(section.get("bears", ""))
-        section_local_bears, section_global_bears = collect_bears(
-            bear_dirs,
-            bears,
-            [BEAR_KIND.LOCAL, BEAR_KIND.GLOBAL],
-            log_printer)
-        all_bears = copy.deepcopy(section_local_bears)
-        all_bears.extend(section_global_bears)
-        fill_section(section, acquire_settings, log_printer, all_bears)
+        bear_dirs_by_section[section_name] = section.bear_dirs()
+        bears_by_section[section_name] = list(section.get("bears", ""))
+
+    bears_to_collect = set(chain.from_iterable(bears_by_section.values()))
+    bear_dirs_to_collect = set(chain(*bear_dirs_by_section.values()))
+
+    all_local_bears, all_global_bears = collect_bears(
+        bear_dirs_to_collect,
+        bears_to_collect,
+        [BEAR_KIND.LOCAL, BEAR_KIND.GLOBAL],
+        log_printer)
+    all_bears = copy.deepcopy(all_local_bears)
+    all_bears.extend(all_global_bears)
+
+    for section_name, section in sections.items():
+        section_bears = []
+        section_local_bears = []
+        section_global_bears = []
+
+        for bear in all_bears:
+            if bear.name in bears_by_section[section_name]:
+                section_bears.append(bear)
+
+                if bear.kind() == BEAR_KIND.LOCAL:
+                    section_local_bears.append(bear)
+                elif bear.kind() == BEAR_KIND.GLOBAL:
```

Comment:
the coverage problem here is that the elif will always be true. Just make an `else` instead of `elif` as it can't be anything else, right?